File: app/src/main/java/org/alkaline/taskbrain/data/SignalListener.kt
package org.alkaline.taskbrain.data

import android.util.Log
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration

/**
 * Single shared snapshot listener on `users/{uid}` that demultiplexes per-
 * channel signal changes to subscribed callbacks. Replaces what used to be
 * one `addSnapshotListener` per subsystem (NoteStore + AlarmRepository),
 * collapsing the bookkeeping into one target.
 *
 * Each [UserDocSignal.Channel] subscriber gets fired when *its* field on
 * the user doc transitions to a new value. Fields owned by other channels
 * are ignored by that subscriber's dedup.
 *
 * Lifecycle:
 * - [attach] runs from whichever subsystem first needs the listener. Safe
 *   to call repeatedly; subsequent calls are no-ops if the listener is up.
 * - [subscribe] returns an unsubscribe function. Subsystems call it from
 *   their `clear` (logout) paths.
 * - [clear] removes the underlying listener and drops all subscribers.
 *   Call on logout from the same place that resets the other repos.
 */
object SignalListener {
    private const val TAG = "SignalListener"

    private val lock = Any()
    private var registration: ListenerRegistration? = null
    private val subscribers = mutableMapOf<UserDocSignal.Channel, MutableList<() -> Unit>>()
    private val lastObserved = mutableMapOf<UserDocSignal.Channel, Timestamp?>()

    /** Idempotent. Attaches the user-doc snapshot listener. */
    fun attach(db: FirebaseFirestore, userId: String) {
        synchronized(lock) {
            if (registration != null) return
            registration = db.collection("users").document(userId)
                .addSnapshotListener { snapshot, error ->
                    if (error != null) {
                        Log.e(TAG, "signal listener error for userId=$userId", error)
                        return@addSnapshotListener
                    }
                    if (snapshot == null) return@addSnapshotListener
                    if (snapshot.metadata.hasPendingWrites()) {
                        // Our own bump's local echo; the server-confirmed
                        // delivery follows.
                        FirestoreUsage.recordRead(
                            "SignalListener",
                            FirestoreUsage.ReadType.LISTENER_LOCAL_ECHO,
                            1,
                        )
                        return@addSnapshotListener
                    }
                    val fromCache = snapshot.metadata.isFromCache
                    FirestoreUsage.recordRead(
                        "SignalListener",
                        if (fromCache) FirestoreUsage.ReadType.LISTENER_UPDATE_CACHED
                        else FirestoreUsage.ReadType.LISTENER_UPDATE_FRESH,
                        1,
                    )
                    dispatch(snapshot)
                }
        }
    }

    private fun dispatch(snapshot: com.google.firebase.firestore.DocumentSnapshot) {
        // Capture firing callbacks under the lock so we don't hold it while
        // user code runs (which may itself try to subscribe/unsubscribe).
        val toFire = mutableListOf<() -> Unit>()
        synchronized(lock) {
            for (channel in UserDocSignal.Channel.values()) {
                val newSignal = snapshot.getTimestamp(channel.fieldName)
                if (newSignal == lastObserved[channel]) continue
                lastObserved[channel] = newSignal
                subscribers[channel]?.let { toFire.addAll(it) }
            }
        }
        for (cb in toFire) {
            try {
                cb()
            } catch (e: Throwable) {
                Log.e(TAG, "subscriber callback threw", e)
            }
        }
    }

    /** Register a callback for [channel]. Returns an unsubscribe function.
     *  The callback fires on every subsequent change to the channel's field;
     *  it does NOT fire retroactively for the most recent value, so callers
     *  that need initial-state behavior should perform their own initial
     *  read/pull on subscribe. */
    fun subscribe(channel: UserDocSignal.Channel, onChange: () -> Unit): () -> Unit {
        synchronized(lock) {
            subscribers.getOrPut(channel) { mutableListOf() }.add(onChange)
        }
        return {
            synchronized(lock) {
                subscribers[channel]?.remove(onChange)
            }
        }
    }

    /** Detach the listener and drop all subscribers. Call on sign-out. */
    fun clear() {
        synchronized(lock) {
            registration?.remove()
            registration = null
            subscribers.clear()
            lastObserved.clear()
        }
    }
}

File: web/src/data/SignalListener.ts
import {
  doc,
  onSnapshot,
  type Firestore,
  type DocumentSnapshot,
  type Unsubscribe,
} from 'firebase/firestore'
import type { Timestamp } from 'firebase/firestore'
import { firestoreUsage } from './FirestoreUsage'
import { type Channel, channelFieldName } from './UserDocSignal'

/**
 * Single shared snapshot listener on `users/{uid}` that demultiplexes per-
 * channel signal changes to subscribed callbacks. Mirrors the Android
 * SignalListener. Each [Channel] subscriber fires when *its* field on the
 * user doc transitions to a new value; fields owned by other channels are
 * ignored by that subscriber's dedup.
 */
class SignalListenerImpl {
  private unsubscribe: Unsubscribe | null = null
  private subscribers = new Map<Channel, Set<() => void>>()
  private lastObserved = new Map<Channel, Timestamp | null>()

  /** Idempotent. Attaches the user-doc snapshot listener. */
  attach(db: Firestore, userId: string): void {
    if (this.unsubscribe) return
    const ref = doc(db, 'users', userId)
    this.unsubscribe = onSnapshot(
      ref,
      (snapshot) => {
        if (snapshot.metadata.hasPendingWrites) {
          firestoreUsage.recordRead('SignalListener', 'LISTENER_LOCAL_ECHO', 1)
          return
        }
        firestoreUsage.recordRead(
          'SignalListener',
          snapshot.metadata.fromCache ? 'LISTENER_UPDATE_CACHED' : 'LISTENER_UPDATE_FRESH',
          1,
        )
        this.dispatch(snapshot)
      },
      (error) => {
        console.error('[SignalListener] error', error)
      },
    )
  }

  private dispatch(snapshot: DocumentSnapshot): void {
    const toFire: Array<() => void> = []
    const data = snapshot.data() ?? {}
    for (const channel of channels()) {
      const newSignal = (data[channelFieldName(channel)] as Timestamp | undefined) ?? null
      if (timestampsEqual(newSignal, this.lastObserved.get(channel) ?? null)) continue
      this.lastObserved.set(channel, newSignal)
      const subs = this.subscribers.get(channel)
      if (subs) for (const cb of subs) toFire.push(cb)
    }
    // Fire callbacks outside the lock-equivalent — JS is single-threaded but
    // a callback might re-enter subscribe/unsubscribe, and iterating the Set
    // while it mutates would be undefined behaviour.
    for (const cb of toFire) {
      try {
        cb()
      } catch (e) {
        console.error('[SignalListener] subscriber callback threw', e)
      }
    }
  }

  /** Register a callback. Returns an unsubscribe function. Does not fire
   *  retroactively — callers needing initial-state behaviour should kick
   *  off their own initial read/pull on subscribe. */
  subscribe(channel: Channel, onChange: () => void): () => void {
    let set = this.subscribers.get(channel)
    if (!set) {
      set = new Set()
      this.subscribers.set(channel, set)
    }
    set.add(onChange)
    return () => {
      this.subscribers.get(channel)?.delete(onChange)
    }
  }

  /** Detach the listener and drop all subscribers. Call on sign-out / detach. */
  clear(): void {
    this.unsubscribe?.()
    this.unsubscribe = null
    this.subscribers.clear()
    this.lastObserved.clear()
  }
}

function channels(): Channel[] {
  return ['NOTES', 'ALARMS']
}

function timestampsEqual(a: Timestamp | null, b: Timestamp | null): boolean {
  if (a === b) return true
  if (a == null || b == null) return false
  return a.seconds === b.seconds && a.nanoseconds === b.nanoseconds
}

export const SignalListener = new SignalListenerImpl()
